cdap-app-fabric/src/main/java/co/cask/cdap/app/stream/DefaultStreamWriter.java

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright © 2015 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.app.stream;
+
+import co.cask.cdap.api.common.Bytes;
+import co.cask.cdap.api.data.stream.StreamBatchWriter;
+import co.cask.cdap.api.data.stream.StreamWriter;
+import co.cask.cdap.api.stream.StreamEventData;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.common.discovery.RandomEndpointStrategy;
+import co.cask.cdap.proto.Id;
+import co.cask.common.http.HttpMethod;
+import co.cask.common.http.HttpRequest;
+import co.cask.common.http.HttpRequests;
+import co.cask.common.http.HttpResponse;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Maps;
+import com.google.common.net.HttpHeaders;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import org.apache.twill.discovery.Discoverable;
+import org.apache.twill.discovery.DiscoveryServiceClient;
+import org.apache.twill.discovery.ServiceDiscovered;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.HttpURLConnection;
+import java.net.InetSocketAddress;
+import java.net.URL;
+import java.nio.ByteBuffer;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Default implementation of {@link StreamWriter}
+ */
+public class DefaultStreamWriter implements StreamWriter {
+
+  private static final Logger LOG = LoggerFactory.getLogger(DefaultStreamWriter.class);
+
+  private final String namespaceId;
+  private final DiscoveryServiceClient discoveryServiceClient;
+
+  public DefaultStreamWriter(String namespaceId, DiscoveryServiceClient discoveryServiceClient) {
+    this.namespaceId = namespaceId;
+    this.discoveryServiceClient = discoveryServiceClient;
+  }
+
+  private URL getStreamURL(String stream) throws IOException {
+    return getStreamURL(stream, false);
+  }
+
+  private URL getStreamURL(String stream, boolean batch) throws IOException {
+    ServiceDiscovered discovered = discoveryServiceClient.discover(Constants.Service.STREAMS);
+    Discoverable discoverable = new RandomEndpointStrategy(discovered).pick(1, TimeUnit.SECONDS);
```

Comment:
You don't need to new an EndpointStrategy every time. You can create once in constructor and keep using it.